### ISSUE TITLE
Add support for managing worktrees

### ISSUE THREAD

**Issue by CheyenneWills** [sublimehq/sublime_merge#757]
**Problem description**

There are no direct facilities (tools, menus or keybindings) for managing worktrees. 

It appears that the SM is already aware of worktrees in that a worktree shows in the "Locations" panel of a repository, but there is not an indicator that it is a worktree)

**Preferred solution**
Provide basic tools for performing worktree operations.

- add
- move
- lock
- unlock
- prune
- remove

In the "Locations" panel have some indicator that an "object" is a worktree

Add the ability to open a worktree in a new tab as an right click option on the "object" from the "Locations" panel.

**Comment by themilkman** [comment]
Actually I might missunderstand @CheyenneWills part '..a worktree shows in the "Locations" panel...' because I don't see it at all. But generally an integration of worktrees would be sweet.

**Comment by teameh** [comment]
Good support for git worktrees would be really great feature! Als would also be a unique selling point for Sublime Merge! 👍🏼

**Comment by tomasorti** [comment]
Yes, as a holder of a license of Sublime Merge (SM), I became a bit disappointed to realize that it does not handle worktrees.
As @themilkman mentioned, I do not see anything that shows that SM recognizes a worktree.
I also commented it in SM forum, someone told it did support it, which is not true.

A reference a found:
https://forum.sublimetext.com/t/sublime-merge-cannot-open-git-worktree/39096

The item I posted:
https://forum.sublimetext.com/t/how-does-worktree-work-in-sublime-merge/59011

**Comment by themilkman** [comment]
Well, you can actually work with worktrees, but opening/creating/deleting is not supported via UI.

Generally I am also not sure how "stable" git worktrees are considered by [git itself](https://git-scm.com/docs/git-worktree#_bugs).

Moreover I have the impression that some users are confused by the official git wording and think of the currently checked out repository, not the actual "new" git worktree feature when writing. But I could be wrong on that.

**Comment by tomasorti** [comment]
What do you mean by "_you can actually work with worktrees_"?
I could work with the worktree I created, but **using git commands on the terminal not via SM**.
SM didn't recognize the worktree, neither helped me with the files diffs (I know, I know, UI is not supported).
So I can't see how SM _can actually_ work with worktrees or help with them... It basically gives zero help.
BTW, my environment is WSL on Windows 10.

**Comment by themilkman** [comment]
Ouh okay, for me it works as any other git repository (aka I can do anything inside, only no UI). Kubuntu 21.04 here.

**Comment by tomasorti** [comment]
I was having issues because I develop under WSL and was using SM for Windows.
If I use SM for WSL (from Linux apt repo) it works.
SM handles worktrees opening them in a different tab.

I would prefer SM for Windows to handle also WSL stuff, since I prefer to use OS native tools.
But I get that would need extra work to handle WSL paths and all that stuff.

**Comment by ckaminski** [comment]
Using worktrees to keep several projects running in a large repo where switching branches takes 3-4 minutes is beneficial.  I use SM all the time, been licensed since day 1.  I'd love better support for this feature, even if it's just opening that worktree branch in a new tab.

**Comment by samholmes** [comment]
👍 

I use SM as my daily driver. I use VSCode and GitLens+ has support for worktrees. Now that I created a worktree for master and other branches, I am getting the error described over here: https://github.com/sublimehq/sublime_merge/issues/1347 Can't SM detect if a git repo is using worktrees and then know to switch the branch pointer or something? I'd love to not have to dance around this incompatibility. Also, I'm a paid user.

**Comment by maddanio** [comment]
neovim anyone? :-D

**Comment by ckaminski** [comment]
I notice in the latest build that in my repo with submodules that if I click on said submodule SM will open that repo in a new tab.

A simple fix to add support would be to a worktree section in the branches sidebar, and double clicking said worktree would open that worktree in a new tab.

Then some repository menu options or right click options to manage worktreees.

This is the last major feature I need the git-cli for.

**Comment by CheyenneWills** [comment]
> I notice in the latest build that in my repo with submodules that if I click on said submodule SM will open that repo in a new tab.
> 
> A simple fix to add support would be to a worktree section in the branches sidebar, and double clicking said worktree would open that worktree in a new tab.
> 
> Then some repository menu options or right click options to manage worktreees.
> 
> This is the last major feature I need the git-cli for.

What I see lacking still is the ability to manage worktrees (e.g. creating them, removing them, etc.).  If I manually create a worktree, SM is able to use the worktree just like any other git repo.

**Comment by themilkman** [comment]
Just ftr, Tower now has worktree support, too with (according to their screenshots) a simple and useful integration.

**Comment by mcarlson** [comment]
Hmmm.... Might have to go back to Tower - which is unfortunate because I love Sublime Merge and just re-upped for both it and Sublime Text :/

**Comment by OranguTech** [comment]
Worktree support is no longer a new/obscure feature, it's practically table-stakes at this point.

**Comment by CheyenneWills** [comment]
Just to update the ticket.

What this request really is for is for the awareness and management of git worktrees.

If I manually create a worktree and point sublime_merge at it as if it's just another repo, it works.  I can check out branches, make commits, etc.. 

What is lacking is the awareness that a repository is a worktree and facilities to perform worktree related operations from within sublime_merge.

**Comment by yaroslavyaroslav** [comment]
Command to create worktrees through branch picker. `linux_repo:feature/some_feature`  -- > `linux_repo-some_feature`.


`~/Library/Application Support/Sublime Merge/Packages/User/Default.sublime-commands`
```json
  {
    "caption": "Worktree: Create… (../<name>) and open",
    "command": "git",
    "args": {
      "argv": [
        "-c",
        "alias.wtnew=!~/Library/Application\\ Support/Sublime\\ Merge/Packages/User/worktree_create_and_open.sh",
        "wtnew",
        "$select_branch"
      ],
    },
  },
```

`worktree_create_and_open.sh `

```bash
#!/bin/sh
set -eu

input_ref="${1-}"
if [ -z "$input_ref" ]; then
  echo "Usage: worktree_create_and_open.sh <branch>" >&2
  exit 2
fi

if ! git rev-parse --is-inside-work-tree >/dev/null 2>&1; then
  echo "Error: not inside a git repository" >&2
  exit 2
fi

repo_root="$(git rev-parse --show-toplevel)"
repo_name="${repo_root##*/}"
repo_name_sanitized="$(printf %s "$repo_name" | tr ' /' '--')"

# Normalize what Sublime Merge might pass (local branch, remote branch, or refs/*).
# Goal:
# - worktree path: ../<last path component>  (feature/some -> ../some)
# - local branch name: strip remote prefix (origin/feature/some -> feature/some)
branch="$input_ref"
start_point="$input_ref"
case "$input_ref" in
  refs/heads/*)
    branch="${input_ref#refs/heads/}"
    start_point="$branch"
    ;;
  refs/remotes/*)
    # refs/remotes/origin/feature/some -> origin/feature/some
    remote_ref="${input_ref#refs/remotes/}"
    branch="${remote_ref#*/}"
    start_point="$remote_ref"
    ;;
  remotes/*)
    # remotes/origin/feature/some -> origin/feature/some
    remote_ref="${input_ref#remotes/}"
    branch="${remote_ref#*/}"
    start_point="$remote_ref"
    ;;
  origin/*)
    branch="${input_ref#origin/}"
    start_point="$input_ref"
    ;;
esac

name="${branch##*/}"
dir="../${repo_name_sanitized}-${name}"

# If the folder already exists, just open it.
if [ -e "$dir" ]; then
  smerge "$dir" >/dev/null 2>&1 || true
  exit 0
fi

if git show-ref --verify --quiet "refs/heads/$branch"; then
  git worktree add "$dir" "$branch"
else
  if [ "$start_point" = "$branch" ]; then
    git worktree add -b "$branch" "$dir"
  else
    git worktree add -b "$branch" "$dir" "$start_point"
  fi
fi

smerge --background "$dir" >/dev/null 2>&1 || true
```

Disclaimer: code is written by AI w/o thorough review.

**Comment by dpjohnst** [comment]
Hi all,

We have added support for managing worktrees in Sublime Merge Build 2126 (Dev build).

I'd love your feedback on how it fits into your workflow and any improvements you'd like to see.

Cheers,
Dylan

**Comment by samholmes** [comment]
@dpjohnst can you link to the announcement of the feature adding worktrees so we get a good idea on how it works in sublime merge. Better yet a video walkthrough of the feature in the product would help. Thanks!

**Comment by CheyenneWills** [comment]
Looks great :)

It appears that all the features that I wanted are there.

Adding/creating a new worktree was a little confusing on the actual mechanics of what to do with the file dialog.  An example in the doc would help.

I'll let others chime in, but from my standpoint I would say this ticket could be closed.

**Comment by yaroslavyaroslav** [comment]
I literally purchased a bundle just bc of this feature release 😅

**Comment by john-unger** [comment]
Lately I have been needing worktree support that I was going to file a request, come to find out this is getting ready 🎉 
Looking forward to seeing this in the next release!! tysm

**Comment by dpjohnst** [comment]
Hi all,

This has been added in Sublime Merge Build 2130!

Please feel free to leave any feedback here, or raise a new report on the tracker for any larger feedback.

Cheers,
Dylan from Sublime HQ
